src/pkcs11_engine.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include <boost/format.hpp>
+#include <boost/scoped_ptr.hpp>
+
+#include "mococrw/pkcs11_engine.h"
+
+namespace mococrw {
+using namespace openssl;
+
+const std::string Pkcs11Engine::ENGINE_ID = "pkcs11";
+
+Pkcs11Engine::Pkcs11Engine(const std::string &module_path,
+		const std::string &pin) {
+	_engine = _ENGINE_by_id(ENGINE_ID.c_str());
+
+	// Set module path:
+	std::vector<char> module_path_copy(module_path.begin(), module_path.end());
+	module_path_copy.push_back('\0');
+	_ENGINE_ctrl_cmd(_engine.get(), "MODULE_PATH", 0,
+			(void*) &module_path_copy[0], nullptr, 1);
+
+	// Set pin:
+	std::vector<char> pin_copy(pin.begin(), pin.end());
+	pin_copy.push_back('\0');
+	_ENGINE_ctrl_cmd(_engine.get(), "PIN", 0, (void*) &pin_copy[0],
+			nullptr, 1);
+
+	_ENGINE_init(_engine.get());
+}
+
+ENGINE* Pkcs11Engine::internal() {
+	return _engine.get();
+}
+
+const ENGINE* Pkcs11Engine::internal() const {
+	return _engine.get();
+}
+
+AsymmetricPublicKey Pkcs11Engine::loadPublicKey(const std::string& key_id)
+{
+    SSL_EVP_PKEY_Ptr evp_key(_ENGINE_load_public_key(_engine.get(), key_id.c_str()));
+    return AsymmetricPublicKey{std::move(evp_key)};
```

Comment:
You could probably just return it. Maybe needs an explicit constructor call but no need for a temporary object. This then also enables RVO.

Reply:
Done.